File: src/windows/FocusedAgentActivator.cs
using System;
using System.Collections.Generic;

namespace CodexHalo
{
    public static class FocusedAgentActivator
    {
        public static void Activate(
            AgentKind focusedAgent,
            IList<SessionSnapshot> visibleSessions,
            IList<SessionSnapshot> hookSnapshots,
            bool paused,
            FocusedSessionLiveEvidence evidence,
            IDictionary<int, HostProcessRecord> processes,
            int selfProcessId,
            Action activateCodex,
            Action<int> activateHost)
        {
            if (focusedAgent == AgentKind.Codex)
            {
                if (activateCodex != null)
                {
                    activateCodex();
                }
                return;
            }
            int pid = FocusedSessionHostResolver.ResolveProcessId(
                focusedAgent, visibleSessions, hookSnapshots, paused, evidence);
            if (pid <= 0)
            {
                return;
            }
            int hostPid = ProcessTreeHostWalker.ResolveHost(
                pid, processes, selfProcessId);
            if (hostPid <= 0 || activateHost == null)
            {
                return;
            }
            activateHost(hostPid);
        }
    }
}

File: src/macos/Sources/AgentHaloCore/FocusedAgentActivator.swift
import Foundation

public enum FocusedAgentActivator {
    public static func activate(
        focusedAgent: AgentKind,
        visibleSessions: [SessionSnapshot],
        hookSnapshots: [SessionSnapshot],
        paused: Bool,
        evidence: FocusedSessionLiveEvidence,
        processes: [Int32: HostProcessRecord],
        selfProcessId: Int32,
        activateCodex: () -> Void,
        activateHost: (Int32) -> Void
    ) {
        if focusedAgent == .codex {
            activateCodex()
            return
        }
        guard let pid = FocusedSessionHostResolver.resolveProcessId(
            focusedAgent: focusedAgent,
            visibleSessions: visibleSessions,
            hookSnapshots: hookSnapshots,
            paused: paused,
            evidence: evidence
        ) else {
            return
        }
        guard let hostPid = ProcessTreeHostWalker.resolveHost(
            startingProcessId: pid,
            processes: processes,
            selfProcessId: selfProcessId
        ) else {
            return
        }
        activateHost(hostPid)
    }
}
